File: Xamarin_Forms/Xamarin_Forms/GertecXamarinForms/GertecXamarinForms.Android/SAT/PagesSat/Teste.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Java.Lang;
using Android.Widget;
using GertecXamarinForms.Controls.Sat;
using GertecXamarinForms.Droid.Sat;
using GertecXamarinForms.Droid.SAT;
using GertecXamarinForms.Droid.SAT.ServiceSat;
using Android.Text;

[assembly: Xamarin.Forms.Dependency(typeof(Teste))]
namespace GertecXamarinForms.Droid.SAT
{
    [Activity(Label = "Teste")]
    public class Teste : Activity, ISatTestes
    {
        public static string ultimaChaveVenda;
        private AlertDialog alerta;
        private string codigoSessao;

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);

            // Create your application here
        }

        public void cancelarVenda(string codAtivacao, string dadosVenda)
        {
            AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.mContext);
            builder.SetTitle("Atenção");
            builder.SetMessage("Digite a chave de cancelamento");

            EditText inputChaveCancelamento = new EditText(MainActivity.mContext);
            inputChaveCancelamento.Text = GlobalValues.ultimaChaveVenda;
            builder.SetView(inputChaveCancelamento);
            builder.SetNeutralButton("OK", OkAction);

            void OkAction(object sender, DialogClickEventArgs e)
            {
                ultimaChaveVenda = inputChaveCancelamento.Text;
                var codigoAtivacao = codAtivacao.ToString();

                if (!SatUtils.VerificaCodigoAtivacao(codigoAtivacao))
                {
                    SatUtils.MostrarToast(MainActivity.mContext, "Código de Ativação deve ter entre 8 a 32 caracteres!");
                    return;
                }

                if (string.IsNullOrEmpty(ultimaChaveVenda))
                {
                    SatUtils.MostrarToast(MainActivity.mContext, "Digite uma chave de cancelamento!");
                    return;
                }

                repostaSat("CancelarUltimaVenda", codigoAtivacao, dadosVenda);
            }

            alerta = builder.Create();
            alerta.Show();
        }

        public void consultarSat()
        {
            repostaSat("ConsultarSat", GlobalValues.codigoAtivacao.ToString());
        }

        public void enviarVenda(string codAtivacao, string dadosVenda)
        {
            var codigoAtivacao = codAtivacao.ToString();
            if (!SatUtils.VerificaCodigoAtivacao(codigoAtivacao))
            {
                SatUtils.MostrarToast(MainActivity.mContext, "Código de Ativação deve ter entre 8 a 32 caracteres!");
                return;
            }

            repostaSat("EnviarTesteVendas", codigoAtivacao, dadosVenda);
        }

        public void secaoSat(string codAtivacao)
        {
            AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.mContext);
            builder.SetTitle("Atenção");
            builder.SetMessage("Digite o número da sessão");
            EditText inputCodigoSessao = new EditText(MainActivity.mContext);
            inputCodigoSessao.SetRawInputType(InputTypes.ClassNumber);
            builder.SetView(inputCodigoSessao);
            builder.SetNeutralButton("OK", OkAction);
            inputCodigoSessao.Text = GlobalValues.ultimaSessao;

            void OkAction(object sender, DialogClickEventArgs e)
            {
                codigoSessao = inputCodigoSessao.Text.Trim();
                var codigoAtivacao = codAtivacao.ToString();

                if (!SatUtils.VerificaCodigoAtivacao(codigoAtivacao))
                {
                    SatUtils.MostrarToast(MainActivity.mContext, "Código de Ativação deve ter entre 8 a 32 caracteres!");
                    return;
                }

                if (string.IsNullOrEmpty(codigoSessao))
                {
                    SatUtils.MostrarToast(MainActivity.mContext, "Digite um número de sessão!");
                    return;
                }

                repostaSat("ConsultarNumeroSessao", codigoAtivacao);
            }
            alerta = builder.Create();
            alerta.Show();
        }

        public void statusSat(string codAtivacao)
        {
            var codigoAtivacao = codAtivacao.ToString();
            if (!SatUtils.VerificaCodigoAtivacao(codigoAtivacao))
            {
                SatUtils.MostrarToast(MainActivity.mContext, "Código de Ativação deve ter entre 8 a 32 caracteres!");
                return;
            }

            repostaSat("ConsultarStatusOperacional", codigoAtivacao);
        }

        public void testeSat(string codAtivacao, string dadosVenda)
        {
            var codigoAtivacao = codAtivacao.ToString();

            if (!SatUtils.VerificaCodigoAtivacao(codigoAtivacao))
            {
                SatUtils.MostrarToast(MainActivity.mContext, "Código de Ativação deve ter entre 8 a 32 caracteres!");
                return;
            }

            repostaSat("EnviarTesteFim", codigoAtivacao, dadosVenda);
        }

        public void repostaSat(string operacao, string txtCodAtivacao, string dadosVenda = "")
        {
            string retornoOperacao = "";
            switch (operacao)
            {
                case "ConsultarSat":
                    retornoOperacao = MainActivity.satFunctions.ConsultarSat(SatUtils.GerarNumeroSessao);
                    break;
                case "ConsultarStatusOperacional":
                    retornoOperacao = MainActivity.satFunctions.ConsultarStatusOperacional(SatUtils.GerarNumeroSessao, txtCodAtivacao);
                    break;
                case "EnviarTesteFim":
                    retornoOperacao = MainActivity.satFunctions.EnviarTesteFim(txtCodAtivacao, dadosVenda, SatUtils.GerarNumeroSessao);
                    break;
                case "EnviarTesteVendas":
                    retornoOperacao = MainActivity.satFunctions.EnviarTesteVendas(txtCodAtivacao, dadosVenda, SatUtils.GerarNumeroSessao);
                    break;
                case "CancelarUltimaVenda":
                    retornoOperacao = MainActivity.satFunctions.CancelarUltimaVenda(txtCodAtivacao, ultimaChaveVenda, dadosVenda, SatUtils.GerarNumeroSessao);
                    break;
                case "ConsultarNumeroSessao":
                    retornoOperacao = MainActivity.satFunctions.ConsultarNumeroSessao(txtCodAtivacao, Integer.ParseInt(codigoSessao), SatUtils.GerarNumeroSessao);
                    break;
                default:
                    retornoOperacao = "";
                    break;
            }


            RetornoSat retornoSat = OperacaoSat.invocarOperacaoSat(operacao, retornoOperacao);
            /*
             * Está verificação(abaixo) tem como objetivo capturar a "Chave de Consulta" retornado na operação EnviarTesteVendas
             * O valor é armazenado em uma variavel global e quando o usuario abre a tela para cancelar venda, o campo (Chave de Cancelamento) já fica preenchido
             */
            if (operacao == "EnviarTesteVendas")
            {
                //chave ultima venda
                GlobalValues.ultimaChaveVenda = retornoSat.ChaveConsulta;
            }

            //GlobalValues.codigoAtivacao = txtCodAtivacao;
            //* Está função [OperacaoSat.formataRetornoSat] recebe como parâmetro a operação realizada e um objeto do tipo RetornoSat
            //* Retorna uma String com os valores obtidos do retorno da Operação já formatados e prontos para serem exibidos na tela
            // Recomenda-se acessar a função e entender como ela funciona
            string retornoFormatado = OperacaoSat.formataRetornoSat(retornoSat);
            SatUtils.DialogoRetorno(MainActivity.mContext, retornoFormatado);
        }

    }
}

File: Android_Nativo_Java/Android/app/src/main/java/com/gertec/exemplosgertec/ExemploSAT/SatPages/Teste.java
package com.gertec.exemplosgertec.ExemploSAT.SatPages;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.text.InputType;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.gertec.exemplosgertec.Components.ComponentsGertec;
import com.gertec.exemplosgertec.ExemploSAT.ServiceSAT.OperacaoSat;
import com.gertec.exemplosgertec.ExemploSAT.ServiceSAT.RetornoSat;
import com.gertec.exemplosgertec.ExemploSAT.ServiceSAT.SatFunctions;
import com.gertec.exemplosgertec.GlobalValues;
import com.gertec.exemplosgertec.R;

import java.util.Random;

/**
 * Created by jmiranda on 10/12/2015.
 */
public class Teste extends Activity {
    private Button button_consultar, button_status, button_teste, button_vendas, button_cancelamento, button_sessao;
    private EditText txtCodAtivacao;
    private AlertDialog alerta;
    Random gerador = new Random();
    SatFunctions satFunctions;
    public String inputText;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        satFunctions = new SatFunctions(this);
        setContentView(R.layout.teste);
        button_consultar = (Button) findViewById(R.id.buttonConsultar);
        button_status = (Button) findViewById(R.id.buttonStatus);
        button_teste = (Button) findViewById(R.id.buttonFimaFim);
        button_vendas = (Button) findViewById(R.id.buttonTesteVenda);
        button_cancelamento = (Button) findViewById(R.id.buttonCancelamento);
        button_sessao = (Button) findViewById(R.id.buttonConsultarSessao);
        txtCodAtivacao = (EditText) findViewById(R.id.txtCodAtivacao);

        txtCodAtivacao.setText(GlobalValues.codAtivacao);
        button_consultar.setOnClickListener(new View.OnClickListener() {
            public void onClick(View arg0) {
                repostaSat("ConsultarSat");
            }
        });

        button_status.setOnClickListener(new View.OnClickListener() {
            public void onClick(View arg0) {
                if (txtCodAtivacao.getText().toString().length() >= 8) {
                    repostaSat("ConsultarStatusOperacional");
                } else {
                    Toast.makeText(Teste.this, "Código de Ativação deve ter entre 8 a 32 caracteres!", Toast.LENGTH_LONG).show();
                }
            }
        });

        button_teste.setOnClickListener(new View.OnClickListener() {
            public void onClick(View arg0) {
                if (txtCodAtivacao.getText().toString().length() >= 8) {
                    repostaSat("EnviarTesteFim");
                } else {
                    Toast.makeText(Teste.this, "Código de Ativação deve ter entre 8 a 32 caracteres!", Toast.LENGTH_LONG).show();
                }
            }
        });

        button_vendas.setOnClickListener(new View.OnClickListener() {
            public void onClick(View arg0) {
                if (txtCodAtivacao.getText().toString().length() >= 8) {
                    repostaSat("EnviarTesteVendas");
                } else {
                    Toast.makeText(Teste.this, "Código de Ativação deve ter entre 8 a 32 caracteres!", Toast.LENGTH_LONG).show();
                }
            }
        });

        button_cancelamento.setOnClickListener(new View.OnClickListener() {
            public void onClick(View arg0) {
                AlertDialog.Builder builder = new AlertDialog.Builder(Teste.this);
                //define o titulo
                builder.setTitle("Atenção");
                //define a mensagem
                builder.setMessage("Digite a chave de cancelamento");
                EditText input = new EditText(Teste.this);
                input.setText(GlobalValues.ultimaChaveVenda);
                builder.setView(input);
                builder.setNeutralButton("OK", new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int which) {
                                if (input.getText().toString().length() > 0) {
                                    inputText = input.getText().toString();
                                    repostaSat("CancelarUltimaVenda");
                                } else {
                                    Toast.makeText(Teste.this, "Digite uma chave de cancelamento!", Toast.LENGTH_LONG).show();
                                }
                            }
                        }
                );
                //cria o AlertDialog
                alerta = builder.create();
                //Exibe
                alerta.show();
            }
        });

        button_sessao.setOnClickListener(new View.OnClickListener() {
            public void onClick(View arg0) {
                AlertDialog.Builder builder = new AlertDialog.Builder(Teste.this);
                //define o titulo
                builder.setTitle("Atenção");
                //define a mensagem
                builder.setMessage("Digite o número da sessão");
                final EditText input = new EditText(Teste.this);
                input.setInputType(InputType.TYPE_CLASS_NUMBER);
                builder.setView(input);
                builder.setNeutralButton("OK", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                        if (input.getText().toString().length() > 0) {
                            inputText = input.getText().toString();
                            repostaSat("ConsultarNumeroSessao");
                        } else {
                            Toast.makeText(Teste.this, "Digite um número de sessão!", Toast.LENGTH_LONG).show();
                        }
                    }
                });

                //cria o AlertDialog
                alerta = builder.create();
                //Exibe
                alerta.show();
            }
        });
    }

    public void repostaSat(String operacao){
        String retornoOperaca = "";
        switch (operacao){
            case "ConsultarSat":
                retornoOperaca = satFunctions.consultarSat(gerador.nextInt(999999));
                break;
            case "ConsultarStatusOperacional":
                retornoOperaca = satFunctions.consultarStatusOperacional(gerador.nextInt(999999), txtCodAtivacao.getText().toString());
                break;
            case "EnviarTesteFim":
                retornoOperaca = satFunctions.enviarTesteFim(txtCodAtivacao.getText().toString(), gerador.nextInt(999999));
                break;
            case "EnviarTesteVendas":
                retornoOperaca = satFunctions.enviarTesteVendas(txtCodAtivacao.getText().toString(), gerador.nextInt(999999));
                break;
            case "CancelarUltimaVenda":
                retornoOperaca = satFunctions.cancelarUltimaVenda(txtCodAtivacao.getText().toString(), inputText, gerador.nextInt(999999));
                break;
            case "ConsultarNumeroSessao":
                retornoOperaca = satFunctions.consultarNumeroSessao(txtCodAtivacao.getText().toString(), Integer.valueOf(inputText.trim()), gerador.nextInt(999999));
                break;
            default:
                retornoOperaca = "";
                break;
        }
        GlobalValues.codAtivacao = txtCodAtivacao.getText().toString();

        RetornoSat retornoSat = OperacaoSat.invocarOperacaoSat(operacao, retornoOperaca);
        /*
         * Está verificação(abaixo) tem como objetivo capturar a "Chave de Consulta" retornado na operação EnviarTesteVendas
         * O valor é armazenado em uma variavel global e quando o usuario abre a tela para cancelar venda, o campo (Chave de Cancelamento) já fica preenchido
         */
        if (operacao.equals("EnviarTesteVendas")) {
            GlobalValues.ultimaChaveVenda = retornoSat.getChaveConsulta();
//             = GlobalValues.valorCfe;
        }
            //* Está função [OperacaoSat.formataRetornoSat] recebe como parâmetro a operação realizada e um objeto do tipo RetornoSat
        //* Retorna uma String com os valores obtidos do retorno da Operação já formatados e prontos para serem exibidos na tela
        // Recomenda-se acessar a função e entender como ela funciona
        String retornoFormatado = OperacaoSat.formataRetornoSat(retornoSat);
        ComponentsGertec.dialogoRetorno(retornoFormatado, this);
    }
}
